staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go

```diff
@@ -31,63 +29,124 @@ import (
 		namespaceScoped: namespaceScoped,
 		validator: customResourceValidator{
-			namespaceScoped: namespaceScoped,
-			kind:            kind,
-			validator:       validator,
+			namespaceScoped:       namespaceScoped,
+			kind:                  kind,
+			schemaValidator:       schemaValidator,
+			statusSchemaValidator: statusSchemaValidator,
 		},
 	}
 }
 
-func (a customResourceDefinitionStorageStrategy) NamespaceScoped() bool {
+func (a customResourceStrategy) NamespaceScoped() bool {
 	return a.namespaceScoped
 }
 
-func (customResourceDefinitionStorageStrategy) PrepareForCreate(ctx genericapirequest.Context, obj runtime.Object) {
+// PrepareForCreate clears the status of a CustomResource before creation.
+func (customResourceStrategy) PrepareForCreate(ctx genericapirequest.Context, obj runtime.Object) {
+	if utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceSubResources) {
+		customResourceObject := obj.(*unstructured.Unstructured)
+		customResource := customResourceObject.UnstructuredContent()
+
+		// create cannot set status
+		if _, ok := customResource["status"]; ok {
+			delete(customResource, "status")
+		}
+	}
+
+	accessor, _ := meta.Accessor(obj)
+	accessor.SetGeneration(1)
 }
 
-func (customResourceDefinitionStorageStrategy) PrepareForUpdate(ctx genericapirequest.Context, obj, old runtime.Object) {
+// PrepareForUpdate clears fields that are not allowed to be set by end users on update.
+func (customResourceStrategy) PrepareForUpdate(ctx genericapirequest.Context, obj, old runtime.Object) {
+	if utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceSubResources) {
+		newCustomResourceObject := obj.(*unstructured.Unstructured)
+		oldCustomResourceObject := old.(*unstructured.Unstructured)
+
+		newCustomResource := newCustomResourceObject.UnstructuredContent()
+		oldCustomResource := oldCustomResourceObject.UnstructuredContent()
+
+		// update is not allowed to set status
+		_, ok1 := newCustomResource["status"]
+		_, ok2 := oldCustomResource["status"]
+
+		switch {
+		case ok1 && ok2:
+			newCustomResource["status"] = oldCustomResource["status"]
+		case ok1 && !ok2:
+			delete(newCustomResource, "status")
+		}
+
+		// Any changes to the spec increment the generation number, any changes to the
+		// status should reflect the generation number of the corresponding object. We push
+		// the burden of managing the status onto the clients because we can't (in general)
+		// know here what version of spec the writer of the status has seen. It may seem like
+		// we can at first -- since obj contains spec -- but in the future we will probably make
+		// status its own object, and even if we don't, writes may be the result of a
+		// read-update-write loop, so the contents of spec may not actually be the spec that
+		// the CustomResource has *seen*.
+
+		newSpec, ok1 := newCustomResource["spec"]
+		oldSpec, ok2 := oldCustomResource["spec"]
+		if ok1 && ok2 {
```

Comment:
can the spec be deleted? Shouldn't that increase the generation as well? Also the spec could have been nil before and is now added.